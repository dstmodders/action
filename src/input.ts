import * as core from '@actions/core';

export interface Input {
  busted: boolean;
  ignoreFailure: boolean;
  luacheck: boolean;
  prettier: boolean;
  slack: boolean;
  slackLuacheckFormat: string;
  slackPrettierFormat: string;
  slackStyLuaFormat: string;
  stylua: boolean;
}

function getSlackFormat(name: string): string {
  const value = core.getInput(name);
  if (
    value.length > 0 &&
    (value === 'failures' || value === 'issues' || value === 'passes')
  ) {
    return value;
  }
  if (value.length === 0) {
    return 'issues';
  }
  throw new Error(
    `Invalid ${name} input value. Should be: failures|issues|passes`,
  );
}

export async function get(): Promise<Input> {
  try {
    const input: Input = <Input>{};
    input.busted = core.getBooleanInput('busted');
    input.ignoreFailure = core.getBooleanInput('ignore-failure');
    input.luacheck = core.getBooleanInput('luacheck');
    input.prettier = core.getBooleanInput('prettier');
    input.slack = core.getBooleanInput('slack');
    input.slackLuacheckFormat = getSlackFormat('slack-luacheck-format');
    input.slackPrettierFormat = getSlackFormat('slack-prettier-format');
    input.slackStyLuaFormat = getSlackFormat('slack-stylua-format');
    input.stylua = core.getBooleanInput('stylua');
    return input;
  } catch (error) {
    return Promise.reject(error);
  }
}
