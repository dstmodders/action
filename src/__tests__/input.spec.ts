import * as input from '../input';
import { Input } from '../input';

const { expect } = require('expect');

describe('input', () => {
  describe('get()', () => {
    describe('when no inputs are set', () => {
      it('should throw an error', async () => {
        expect.assertions(1);
        try {
          await input.get();
        } catch (error) {
          if (error instanceof Error) {
            expect(error.message).toMatch(
              'Input does not meet YAML 1.2 "Core Schema" specification: busted',
            );
          }
        }
      });
    });

    describe('when all inputs are set', () => {
      describe('as default values', () => {
        beforeEach(() => {
          process.env.INPUT_BUSTED = 'false';
          process.env.INPUT_LDOC = 'false';
          process.env.INPUT_LUACHECK = 'false';
          process.env.INPUT_PRETTIER = 'false';
          process.env.INPUT_SLACK = 'false';
          process.env.INPUT_STYLUA = 'false';
          process.env['INPUT_IGNORE-CHECK-VERSIONS'] = 'false';
          process.env['INPUT_IGNORE-FAILURE'] = 'false';
          process.env['INPUT_IGNORE-SET-OUTPUT'] = 'false';
          process.env['INPUT_SLACK-COLOR-DEFAULT'] = '#1f242b';
          process.env['INPUT_SLACK-COLOR-FAILURE'] = '#cc1f2d';
          process.env['INPUT_SLACK-COLOR-SUCCESS'] = '#24a943';
          process.env['INPUT_SLACK-COLOR-WARNING'] = '#dcad04';
          process.env['INPUT_SLACK-FORCE-STATUS'] = '';
          process.env['INPUT_SLACK-LUACHECK-FORMAT'] = 'issues';
          process.env['INPUT_SLACK-PRETTIER-FORMAT'] = 'issues';
          process.env['INPUT_SLACK-STYLUA-FORMAT'] = 'issues';
        });

        it("shouldn't throw an error", async () => {
          expect.assertions(0);
          try {
            await input.get();
          } catch (error) {
            expect(error).not.toBeNull();
          }
        });

        it('should match the snapshot', async () => {
          expect(await input.get()).toMatchSnapshot();
        });

        describe('and slack-color-default is set', () => {
          describe('and valid', () => {
            beforeEach(() => {
              process.env['INPUT_SLACK-COLOR-DEFAULT'] = '#cccccc';
            });

            it("shouldn't throw an error", async () => {
              expect.assertions(0);
              try {
                await input.get();
              } catch (error) {
                expect(error).not.toBeNull();
              }
            });
          });

          describe('and invalid', () => {
            beforeEach(() => {
              process.env['INPUT_SLACK-COLOR-DEFAULT'] = 'test';
            });

            it('should throw an error', async () => {
              expect.assertions(1);
              try {
                await input.get();
              } catch (error) {
                if (error instanceof Error) {
                  expect(error.message).toMatch(
                    'Invalid slack-color-default input value. Should be a valid HEX color',
                  );
                }
              }
            });
          });
        });

        describe('and slack-force-status is set', () => {
          describe('and valid', () => {
            beforeEach(() => {
              process.env['INPUT_SLACK-FORCE-STATUS'] = 'skipped';
            });

            it("shouldn't throw an error", async () => {
              expect.assertions(0);
              try {
                await input.get();
              } catch (error) {
                expect(error).not.toBeNull();
              }
            });
          });

          describe('and invalid', () => {
            beforeEach(() => {
              process.env['INPUT_SLACK-FORCE-STATUS'] = 'test';
            });

            it('should throw an error', async () => {
              expect.assertions(1);
              try {
                await input.get();
              } catch (error) {
                if (error instanceof Error) {
                  expect(error.message).toMatch(
                    'Invalid slack-force-status input value. Should be: success|failure|cancelled|skipped',
                  );
                }
              }
            });
          });
        });

        describe('and slack-luacheck-format is empty', () => {
          beforeEach(() => {
            process.env['INPUT_SLACK-LUACHECK-FORMAT'] = '';
          });

          it("shouldn't throw an error", async () => {
            expect.assertions(0);
            try {
              await input.get();
            } catch (error) {
              expect(error).not.toBeNull();
            }
          });

          it('should have the "issues" value', async () => {
            const result: Input = await input.get();
            expect(result.slackLuacheckFormat).toMatch('issues');
          });
        });

        describe('and slack-luacheck-format is invalid', () => {
          beforeEach(() => {
            process.env['INPUT_SLACK-LUACHECK-FORMAT'] = 'test';
          });

          it('should throw an error', async () => {
            expect.assertions(1);
            try {
              await input.get();
            } catch (error) {
              if (error instanceof Error) {
                expect(error.message).toMatch(
                  'Invalid slack-luacheck-format input value. Should be: issues|passes|failures',
                );
              }
            }
          });
        });
      });
    });
  });
});
