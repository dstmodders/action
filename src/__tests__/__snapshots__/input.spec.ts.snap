// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input get() when all inputs are set as default values should match the snapshot 1`] = `
Object {
  "busted": false,
  "ignoreCheckVersions": false,
  "ignoreFailure": false,
  "ignoreSetOutput": false,
  "ldoc": false,
  "luacheck": false,
  "prettier": false,
  "slack": false,
  "slackColorDefault": "#1f242b",
  "slackColorFailure": "#cc1f2d",
  "slackColorSuccess": "#24a943",
  "slackColorWarning": "#dcad04",
  "slackForceStatus": "",
  "slackLuacheckFormat": "issues",
  "slackPrettierFormat": "issues",
  "slackStyLuaFormat": "issues",
  "stylua": false,
}
`;
